File: backend/app/schemas/ai.py
from __future__ import annotations

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel


class MoodReportResponse(BaseModel):
    """Response model for mood report."""
    period: str  # "week" | "month"
    summary: str
    emotion_distribution: dict[str, int]  # e.g. {"positive": 5, "neutral": 3, "negative": 1}
    daily_moods: list[dict]  # e.g. [{"date": "2026-07-13", "total": 2, "positive": 1, "neutral": 1, "negative": 0}]
    top_keywords: list[dict]  # e.g. [{"keyword": "生活", "count": 5}]

    class Config:
        from_attributes = True
        json_encoders = {
            date: lambda v: v.isoformat(),
            datetime: lambda v: v.isoformat(),
        }


class ChatRequest(BaseModel):
    """Request model for AI chat."""
    message: str
    context_moment_ids: list[UUID] | None = None


class MoodStatsResponse(BaseModel):
    """Response model for mood statistics (frontend charts)."""
    daily_moods: list[dict]  # [{"date": "2026-07-01", "positive": 3, "neutral": 1, "negative": 0, "count": 4}]
    emotion_pie: dict[str, int]  # {"positive": 45, "neutral": 30, "negative": 15}
    top_tags: list[dict]  # [{"tag": "生活记录", "count": 28}]

    class Config:
        from_attributes = True
        json_encoders = {
            date: lambda v: v.isoformat(),
            datetime: lambda v: v.isoformat(),
        }


class ChatResponse(BaseModel):
    """Response model for AI chat."""
    reply: str

File: frontend/src/services/ai.ts
import { apiClient } from "./client";
import type { ChatMessage, MomentAnalysis, MoodReport } from "../types/api";

// ── Mood Stats types ────────────────────────────────────
export interface DailyMood {
  date: string
  positive: number
  neutral: number
  negative: number
  count: number
}

export interface MoodStats {
  daily_moods: DailyMood[]
  emotion_pie: { positive: number; neutral: number; negative: number }
  top_tags: { tag: string; count: number }[]
}

/**
 * Analyze a single Moment for AI summary / emotion / tags.
 */
export async function analyzeMoment(
  momentId: string
): Promise<MomentAnalysis> {
  return apiClient.post(`/ai/moments/${momentId}/analyze`);
}

/**
 * Fetch mood report for a given period (week / month).
 */
/**
 * Fetch mood statistics (daily breakdown, pie, tags) for the given number of days.
 */
export async function getMoodStats(days: number = 30): Promise<MoodStats> {
  return apiClient.get('/users/me/mood-stats', {
    params: { days },
  }) as unknown as MoodStats;
}

/**
 * Fetch mood report for a given period (week / month).
 */
export async function getMoodReport(params: {
  period: string;
  start_date?: string;
  end_date?: string;
}): Promise<MoodReport> {
  return apiClient.get("/ai/mood-report", { params });
}

/**
 * Send a message to the AI companion chat.
 * Optionally attach context moment IDs for the AI to reference.
 */
export async function chatWithAI(
  message: string,
  contextMomentIds?: string[]
): Promise<ChatMessage> {
  return apiClient.post("/ai/chat", {
    message,
    context_moment_ids: contextMomentIds,
  });
}
